kernel/aster-nix/src/thread/task.rs

```diff
@@ -63,7 +63,7 @@ pub fn create_new_user_task(user_space: Arc<UserSpace>, thread_ref: Weak<Thread>
                 break;
             }
             // a preemption point after handling user event.
-            preempt(current_task);
+            schedule();
```

Comment:
A call to `schedule` is relatively expensive as it at least incurs the overhead of taking a lock. Thus, we do not want to call `schedule` on the critical path of every system call. Instead, add `should_preempt` and `set_should_preempt` methods to `Task` so that we can do

```
if current_task.should_preempt() {
    schedule();
}
```

In addition, the `UserPreemption` in the `arch::x86::cpu` module should also be replaced with the code pattern above.

Don't forget to update the registered timer callback to call `Task::set_should_preempt` when `LocalRunQueue::update_current` returns `true`.

Reply:
I saw the revised implementation of `schedule`, which includes the optimization proposed above. But there is a more fundamental problem: `osdk` should never have provided a `schedule` method in the first place.

As a first sign, the `schedule` API does fall in the same league as other methods provided by the `sched` module such as `enqueue` (spawning task), `dequeue` (exiting task), `yield`, `try_wait`, and `wake_up`. You see, all other methods give the reason for taking any scheduling action. But `schedule` is not. So this `schedule` method is logically inconsistent with other methods.

Second, the call to `schedule` here is to serve as a preemption point.

```rust
            // a preemption point after handling user event.
            schedule();
```

But by definition, preemption does not demand any cooperation from tasks. As a `ostd` user, `aster-nix` should never have been calling a `schedule` cooperatively. Currently, we only do user-mode preemption. It should be done inside `UserMode::execute`.

```rust
let return_reason = user_mode.execute(has_kernel_event_fn);
```

So I suggest removing this public `schedule` from `sched` and add a crate-visibile `preempt` method, which preempts the current task only if the `SHOULD_PREEMPT` flag is set to true.

This is related to my advise of removing `UserPreemption`.

Reply:
We talked about to move the check of kernel events and preemption before `UserContext::run` in the discussion.

In the latest [commit](bc7c29f444e4396d8f5aeb78e42e91d7d9957116), I only moved the preemption part because I can observe a notable drop in performance even in the getpid test which for now I don't know why.